### ISSUE TITLE
Guidelines on merging

### ISSUE THREAD

**Issue by pierrecagne** [UniMath/SymmetryBook#85]
This is not totally an issue but I don't see a way to just make a public comment on an GitHub project.

From time to time, we have a commit of the form ` Merge branch 'master' of https://github.com/UniMath/SymmetryBook`. From my understanding, it happens in the following situation:
  1. person A and person B both have local master up-to-date with origin/master
  2. person A is doing some work on his local master branch
  3. person B is pushing modifications from his/her local master branch to origin/master
  4. person A wants to push his/her modifications also but is now no longer up-to-date
  5. person A fetch origin/master and merge origin/master into his/her local master branch
  6. this results in a merge commit on person A's local master 
  7. person A can now push to origin/master, with this merge commit as the last one

The "problem" resides in step 5. One should not merge origin/master into our master branch to then push master to origin/master, which ends up in a constant changing of the main history of the origin/master branch. This happens when one does `git pull`, which is a shortcut for `git fetch origin && git merge origin/master`. 

Instead, from what I understand, one should rebase  the local master to origin/master: `git fetch origin && git rebase origin/master`. In that way, all of our pending commits will be replayed on top of the origin/master history (if possible, asking for resolving merge conflict if necessary). In other words, it is best practice to think of our own local master as a fork of the project, and pushing modification should be thought as a kind of pull request, hence should be preceeded by a rebase.  It results in a cleaner history for the origin/master branch, and does keep the 'direct parent' links in the intended order.

Long story short: fetch+rebase and fetch+merge both ends up with the same topology for the DAG of commits, but commits history does contain a little more information that just the DAG that some tools actually uses (for graphing the history for example). Using fetch+rebase would allow a newcomer to better follow the commit history of the project.

In any case, this is not an important issue, I just wanted to point out the best practice on this matter (which I actually forget  from times to times, so you will probably see the mistake in my commits also).

**Comment by DanGrayson** [comment]
From the git-pull man page:
```
           See pull.rebase, branch.<name>.rebase and branch.autoSetupRebase in git-config(1) if you
           want to make git pull always use --rebase instead of merging.

               Note
               This is a potentially dangerous mode of operation. It rewrites history, which does
               not bode well when you published that history already. Do not use this option unless
               you have read git-rebase(1) carefully.
```

**Comment by marcbezem** [comment]
As a naive git-user I wondered already why they would offer the command "git pull" as a shorthand if that wasn't the preferred way. But I agree with Pierre to some extent, and have often been confused by seeing my commit under the name of "Merge this or that" joined with many changes that are not mine. Perhaps somebody can explain me what happens in the following case.

Let t1 < t2 < t3 < t4 be time stamps of commits c1,c2,c3,c4, where c1 and c3 are done by B and c2 and c4 by A. Let's assume no conflicts. If B pushes before A, then the history is c1,c3. A is now not up to date.

If A pulls and then pushes, the history becomes c1,c2,c3,c4,m, where m is an extra commit, which includes the commits by B.

If A rebases and then pushes, the history becomes c1,c3,c2,c4, which is not chronological.

Is this what happens?

**Comment by DanGrayson** [comment]
Almost: the history is not linear when there is a merge.

**Comment by marcbezem** [comment]
How does the DAG look in my first case? Perhaps we need an initial c0. Do we then get:
c0 -> c1 -> c3 -> m and c0 -> c2 -> c4 -> m ?

**Comment by DanGrayson** [comment]
In my other software projects I never bother with rebasing, and there are lots of merges.  If I really want to see the structure of a non-linear history, I use "gitk", which gives a good graphical display:

<img width="895" alt="Screen Shot 2020-07-15 at 8 05 39 AM" src="https://user-images.githubusercontent.com/700228/87543045-9f051e80-c669-11ea-92fa-2fd0e4bc19bc.png">

Git log also has options for showing the tree:

```
* | | | | | | 6c49cc1b6c483301e271ddff01ddd88c8a3fd273 Small change to make cmake output more clear (done with Mahrud)
* | | | | | | f9197354d524affd3ed9128d379a72f903d829de added latest version of Complexes.m2, which is planned to replace m2/chaincomplexes.m2 and related files and types at some point.
|/ / / / / /  
* | | | | | 73b3517451d05f708c64d734713fd174dfc689f6 fix expression syntax in workflow
* | | | | | f311a586d478dbf72ea6856672f665543aa7cae2 add "make check" to workflow
* | | | | | 603b1abe73e67821382072169a3dd8bcf27245ef build gmp ...
* | | | | |   cd76dab5b71b030fa530dd045d58d5274ffb97fa Merge pull request #1313 from d-torrance/small-fixes
|\ \ \ \ \ \  
| |_|_|/ / /  
|/| | | | |   
| * | | | | f6206b7617660c016c64dab6aeb3d73bd744727d Fix typo in message that's printed when Macaulay2 is interrupted
| * | | | | 3a722fb2ec08a1eb1c95840af09957d047dbd731 Fix "invalid conversion" error in displayCF debugging function
|/ / / / /  
* | | | | 92af2b930403302538a9311e1b5c400fb4c5e0f9 change a cmake command (in 4ti2 building) from 'rm' to 'remove' so it will still work with cmake 3.16.
* | | | |   8fe72aa8b36786087b0922c6e679a656a9298da1 Merge branch 'development' of https://github.com/Macaulay2/M2 into development
|\ \ \ \ \  
| * \ \ \ \   ba3569b0d060438de0930394ae4dba5462983952 Merge branch 'development' of github.com:Macaulay2/M2 into development
| |\ \ \ \ \  
| * | | | | | b5083b57b19e01cd38fd7ac338bf4829ad6c1e56 fix spelling of variable name
| * | | | | | c151606837874b568efe98aa26011dfa4526f014 resolves #1118
| * | | | | | ee913f6b3462fe3f1bc41b17723830f600149a3e certify three packages published in JSAG
| * | | | | |   db406fea8abd0ac9b76ad2a10c13547464705e1a Merge branch 'pubish-SumsOfSquares' into development
| |\ \ \ \ \ \  
| | * | | | | | 7623915208e2f52445cd73c9acd680027bba1d9c this is the published version of the package SumsOfSquares
| * | | | | | |   15df94c44ae18ee728bec17f00cd6ba1018cc7ad Merge branch 'publish-SemidefiniteProgramming' into development
| |\ \ \ \ \ \ \  
| | * | | | | | | dce05860f624b1f5b58c6b603303c987f7bcc464 this is the published version of SemidefiniteProgramming
| | |/ / / / / /  
| * | | | | | |   a30d7680a956f4991848ccea65b44bd0df5f9e24 Merge branch 'publish-DeterminantalRepresentations' into development
| |\ \ \ \ \ \ \  
| | * | | | | | | f3c4030a3e66ae51f54ec24a89e1d5b1992a82eb this is the published version of the package DeterminantalRepresentations
* | | | | | | | | 4acf8a836d5f09f68a5b33bedcc1d3c3d2b162e2 update PHCpack example output
| |_|_|/ / / / /  
|/| | | | | | |   
* | | | | | | | 5ac7eb2c06c97a473e60cc512b42c1adbe4bacc3 fix spelling of filenames in package NormalToricVariety, part 2
|/ / / / / / /  
* | | | | | | 33f55939ab7e3f0f161895d3bbf5ea6fd097bf5b fix spelling of filename in package NormalToricVariety
* | | | | | |   2145622c046cc8af04507af76c955340df19733f Merge pull request #1296 from mahrud/feature/cmake
|\ \ \ \ \ \ \  
| * \ \ \ \ \ \   f3ed5cf1c9e3df09b8c3a23070892be16c99de95 Merge branch 'development' into feature/cmake
| |\ \ \ \ \ \ \  
| * | | | | | | | b4f567682034b733beddc598f2291b50df522cac temporarily fix Mac issues with shared libnormaliz
| * | | | | | | | 18b296b0c0ba751b93705746835bfefc692a2bb1 temporarily fix Mac issue with lrslib
| * | | | | | | | 95d94b6f56e6aca295c874a62610936ddfaeaa27 finish build-polymake-install
| * | | | | | | | 7ff04d0f839c3ac89613b96e8cbde2b12f99a471 undefined message for old MPSolve
| * | | | | | | | 704b80599b8f5520b37cf855dfa3b739e66495a0 fix ComputationsBook tests in CMake
| * | | | | | | | c785e88a0b7114032a76a1cdbdd0e0656b3c1072 update lrslib, improve normaliz, and fix build-polymake
| * | | | | | | | a24493ac409759b8e87f1b4661d2abfdfd7502d5 add build-phcpack target for v2.4.77 (not built by default)
```

**Comment by DanGrayson** [comment]
> How does the DAG look in my first case? Perhaps we need an initial c0. Do we then get:
> c0 -> c1 -> c3 -> m and c0 -> c2 -> c4 -> m ?

Yes.

**Comment by pierrecagne** [comment]
Yes, I think the description as two paths from c0 to m is accurate Marc. 

Also, `git pull` is perfectly OK when you are sure that origin/master is ahead of your local master. And this is usually how people use git I think: they don't work on the master branch directly, they branch out of the master to add a feature, then merge the new feature in the master branch. So doing a `git pull` while on the master branch is always "safe".  I agree also that the workflow does not match so well a collaborative writing as it does software development.

Again, my initial post is just a statement of what is considered to be best practice from what I found on internet, but we can decide it does not suit our needs.

**Comment by favonia** [comment]
One can use the following command to disable the auto merging, only allowing "fast-forwarding" (that is, it's "safe"):
```sh
git config pull.ff only
```
and then one can decide whether it should be `git merge` or `git rebase`.
